transformer_engine/jax/transformer.py

```diff
@@ -0,0 +1,911 @@
+    return mask
+
+
+def combine_masks(*masks: Optional[Array], dtype: DType = jnp.float32):
+    """Combine attention masks."""
+    func = jnp.logical_and
+    return _merge_mask(func, *masks).astype(dtype)
+
+
+def combine_biases(*masks: Optional[Array]):
+    """Combine attention biases."""
+    func = lambda a, b: a + b
+    return _merge_mask(func, *masks)
+
+
+def core_attention(query: Array,
+                   key: Array,
+                   value: Array,
+                   transpose_batch_sequence: bool,
+                   softmax_type: SoftmaxType = SoftmaxType.SCALED,
+                   softmax_sharding_type: ShardingType = ShardingType.SINGLE,
+                   mask: Optional[Array] = None,
+                   bias: Optional[Array] = None,
+                   dropout_rng: Optional[PRNGKey] = None,
+                   dropout_rate: float = 0.,
+                   deterministic: bool = False,
+                   dtype: DType = jnp.float32,
+                   float32_logits: bool = False):
+    """Core attention"""
+    assert key.ndim == query.ndim == value.ndim, 'q, k, v must have same rank.'
+    batch_dim = 1 if transpose_batch_sequence else 0
+    assert query.shape[batch_dim] == key.shape[batch_dim] == value.shape[batch_dim], (
+        'q, k, v batch dims must match.')
+    assert query.shape[-2] == key.shape[-2] == value.shape[-2], ('q, k, v num_heads must match.')
+    sequence_dim = 0 if transpose_batch_sequence else 1
+    assert key.shape[sequence_dim] == value.shape[sequence_dim], 'k, v lengths must match.'
+    assert query.shape[-1] == key.shape[-1], 'q, k depths must match.'
+
+    if float32_logits:
+        query = query.astype(jnp.float32)
+        key = key.astype(jnp.float32)
+
+    if transpose_batch_sequence:
+        attn_weights = jnp.einsum('qbhd,kbhd->bhqk', query, key)
+    else:
+        attn_weights = jnp.einsum('bqhd,bkhd->bhqk', query, key)
+
+    attn_weights = Softmax(softmax_type=softmax_type,
+                           sharding_type=softmax_sharding_type)(attn_weights, mask, bias)
+
+    if not deterministic and dropout_rate > 0.:
+        keep_prob = 1.0 - dropout_rate
+        dropout_shape = list(attn_weights.shape)
+        dropout_shape[-2] = 1
+        keep = jax_random.bernoulli(dropout_rng, keep_prob, dropout_shape)
+        keep = jnp.broadcast_to(keep, attn_weights.shape)
+        multiplier = (keep.astype(attn_weights.dtype) / jnp.asarray(keep_prob, dtype=dtype))
+        attn_weights = attn_weights * multiplier
+
+    if transpose_batch_sequence:
+        return jnp.einsum('bhqk,kbhd->qbhd', attn_weights, value)
+
+    return jnp.einsum('bhqk,bkhd->bqhd', attn_weights, value)
+
+
+dynamic_vector_slice_in_dim = vmap(lax.dynamic_slice_in_dim, in_axes=(None, 0, None, None))
+
+
+class AttentionType(Enum):
+    """TransformerLayerType."""
+    PADDING = "padding_attention"
+    CAUSAL = "causal_attention"
+
+
+class MultiHeadAttention(nn.Module):
+    # pylint: disable=line-too-long
```

Comment:
I would prefer reformatting the docstring rather than suppressing the Pylint warning.

Reply:
@timmoon10 ,
It should be solved by [b1e5386](https://github.com/NVIDIA/TransformerEngine/pull/54/commits/b1e53864eff0d85c3e956f9e1ddf8948036e46a1)